src/Sylius/Bundle/CoreBundle/spec/Validator/Constraints/ValidCatalogPromotionRulesValidatorSpec.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace spec\Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\CoreBundle\Validator\Constraints\ValidCatalogPromotionRules;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Symfony\Component\Validator\Violation\ConstraintViolationBuilderInterface;
+
+final class ValidCatalogPromotionRulesValidatorSpec extends ObjectBehavior
+{
+    function let(ExecutionContextInterface $executionContext, ProductVariantRepositoryInterface $variantRepository): void
+    {
+        $this->beConstructedWith($variantRepository);
+
+        $this->initialize($executionContext);
+    }
+
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldHaveType(ConstraintValidator::class);
+    }
+
+    function it_adds_violation_if_catalog_promotion_has_invalid_type(
+        ExecutionContextInterface $executionContext,
+        ConstraintViolationBuilderInterface $constraintViolationBuilder,
+        CatalogPromotionRuleInterface $rule
+    ): void {
+        $constraint = new ValidCatalogPromotionRules();
```

Comment:
You can mock this constraint, there is no need in this spec to create a new instance if it